source/common/http/http2/codec_impl.cc

```diff
@@ -747,6 +747,10 @@ ClientConnectionImpl::ClientConnectionImpl(Network::Connection& connection,
 Http::StreamEncoder& ClientConnectionImpl::newStream(StreamDecoder& decoder) {
   StreamImplPtr stream(new ClientStreamImpl(*this, per_stream_buffer_limit_));
   stream->decoder_ = &decoder;
+  // If the connection is currently above the high watermark, make sure to inform the new stream.
+  if (connection_.aboveHighWatermark()) {
```

Comment:
Did you consider doing a similar add/remove callback dance as done at the stream level but at the connection level? Could the approach here have been used at the stream level?